toolchain/semantics/semantics_node.h

```diff
@@ -54,123 +55,108 @@ struct SemanticsNodeBlockId {
-                         SemanticsNodeArgs(one_node));              \
-  }
-#define CARBON_SEMANTICS_MAKE_two_nodes(Name)                          \
-  static auto Make##Name(SemanticsNodeId node1, SemanticsNodeId node2) \
-      ->SemanticsNode {                                                \
-    return SemanticsNode(                                              \
-        SemanticsNodeKind::Name(),                                     \
-        SemanticsNodeArgs(SemanticsTwoNodeIds{node1, node2}));         \
-  }
-
-#define CARBON_SEMANTICS_MAKE_identifier(Name)                              \
-  static auto Make##Name(SemanticsIdentifierId identifier)->SemanticsNode { \
-    return SemanticsNode(SemanticsNodeKind::Name(),                         \
-                         SemanticsNodeArgs(identifier));                    \
-  }
-#define CARBON_SEMANTICS_MAKE_integer_literal(Name)                 \
-  static auto Make##Name(SemanticsIntegerLiteralId integer_literal) \
-      ->SemanticsNode {                                             \
-    return SemanticsNode(SemanticsNodeKind::Name(),                 \
-                         SemanticsNodeArgs(integer_literal));       \
-  }
-#define CARBON_SEMANTICS_MAKE_node_block(Name)                             \
-  static auto Make##Name(SemanticsNodeBlockId node_block)->SemanticsNode { \
-    return SemanticsNode(SemanticsNodeKind::Name(),                        \
-                         SemanticsNodeArgs(node_block));                   \
-  }
-#define CARBON_SEMANTICS_MAKE_node_and_node_block(Name)                      \
-  static auto Make##Name(SemanticsNodeId node,                               \
-                         SemanticsNodeBlockId node_block)                    \
-      ->SemanticsNode {                                                      \
-    return SemanticsNode(                                                    \
-        SemanticsNodeKind::Name(),                                           \
-        SemanticsNodeArgs(SemanticsNodeIdAndNodeBlockId{node, node_block})); \
-  }
-
-#define CARBON_SEMANTICS_NODE_KIND(Name, ArgsType) \
-  CARBON_SEMANTICS_MAKE_##ArgsType(Name)
-#include "toolchain/semantics/semantics_node_kind.def"
-
-#undef CARBON_SEMANTICS_MAKE_no_args
-#undef CARBON_SEMANTICS_MAKE_one_node
-#undef CARBON_SEMANTICS_MAKE_two_nodes
-
-#undef CARBON_SEMANTICS_MAKE_identifier
-#undef CARBON_SEMANTICS_MAKE_integer_literal
-#undef CARBON_SEMANTICS_MAKE_node_block
-#undef CARBON_SEMANTICS_MAKE_node_and_node_block
-
-  SemanticsNode() : kind_(SemanticsNodeKind::Invalid()) {}
+  struct NoArgs {};
+
+  auto GetInvalid() const -> NoArgs { CARBON_FATAL() << "Invalid access"; }
+
+  static auto MakeBinaryOperatorAdd(SemanticsNodeId lhs, SemanticsNodeId rhs)
+      -> SemanticsNode {
+    return SemanticsNode(SemanticsNodeKind::BinaryOperatorAdd(), lhs.id,
+                         rhs.id);
+  }
+  auto GetBinaryOperatorAdd() const
```

Comment:
I found this naming a bit confusing: I expected `SemanticsNode::GetBinaryOperatorAdd` to be something like a factory function, rather than "assume that this is a binary operator add and get its operands". Would something like `DecomposeBinaryOperatorAdd`, or `GetAsBinaryOperatorAdd`, or `GetBinaryOperatorAddArgs` work?

Reply:
Switched to GetAs*

I'd considered Get*Args, but it felt weird to have Args for single-arg things; also I think it's nice to only add on one side. The meaning of Decompose isn't as obvious to me, I suppose?